File: rest/src/main/kotlin/cloud/fabX/fabXaccess/device/rest/Device.kt
package cloud.fabX.fabXaccess.device.rest

import cloud.fabX.fabXaccess.common.rest.ChangeableValue
import kotlinx.serialization.Serializable

@Serializable
data class Device(
    val id: String,
    val aggregateVersion: Long,
    val name: String,
    val background: String,
    val backupBackendUrl: String,
    val attachedTools: Map<Int, String>
)

fun cloud.fabX.fabXaccess.device.model.Device.toRestModel(): Device = Device(
    id = id.serialize(),
    aggregateVersion = aggregateVersion,
    name = name,
    background = background,
    backupBackendUrl = backupBackendUrl,
    attachedTools = attachedTools.mapValues { it.value.serialize() }
)

@Serializable
data class DeviceCreationDetails(
    val name: String,
    val background: String,
    val backupBackendUrl: String,
    val mac: String,
    val secret: String
)

@Serializable
data class DeviceDetails(
    val name: ChangeableValue<String>?,
    val background: ChangeableValue<String>?,
    val backupBackendUrl: ChangeableValue<String>?
)

File: frontend-svelte/src/lib/api/model/device.ts
import type { ChangeableValue } from './changeable-value';
import type { Tool } from '$lib/api/model/tool';

export interface Device {
	id: string;
	aggregateVersion: number;
	name: string;
	background: string;
	backupBackendUrl: string;
	actualFirmwareVersion: string | null;
	desiredFirmwareVersion: string | null;
	attachedTools: Record<number, string>;
}

export interface AugmentedDevice {
	id: string;
	aggregateVersion: number;
	name: string;
	background: string;
	backupBackendUrl: string;
	actualFirmwareVersion: string | null;
	desiredFirmwareVersion: string | null;
	attachedTools: Record<number, Tool>;
}

export interface DeviceCreationDetails {
	name: string;
	background: string;
	backupBackendUrl: string;
	mac: string;
	secret: string;
}

export interface DeviceDetails {
	name: ChangeableValue<string> | null;
	background: ChangeableValue<string> | null;
	backupBackendUrl: ChangeableValue<string> | null;
}

export interface DesiredFirmwareVersion {
	desiredFirmwareVersion: string;
}

export interface ToolAttachmentDetails {
	toolId: string;
}

export interface ToolUnlockDetails {
	toolId: string;
}

export interface AtDeviceCardCreationDetails {
	userId: string;
}
